### ISSUE TITLE
Adding TOME support

### ISSUE THREAD

**Issue by ericboivin** [madarasz/always-be-running#41]
FFG's "official" tournament software is TOME. It's not always simple to manage, but I am sure that with some efforts it could be possible to add a function to import TOME JSON output into ABR. 

Just so you know, I've started working on a similar thing in Java and it is quite a mess. I initially started the reverse engineering in this github repo (in Java) https://github.com/ericboivin/tome2acoo.

I am sure that we could do a joint program and find a way to add TOME import in it ABR. We could maybe use my project as a baseline to create the CSV in the right format so it can be imported in ABR.

The main obstacle to this is the JSON itself. It is a mess! If you or anyone has interest in reverse engineering their structure, we could do something pretty cool, and thus allowing people to push their results on ABR from the "official" tournament reporter!

**Comment by lpoulter** [comment]
This sounds like a really cool idea. Do you have many exported TOME events?

**Comment by lpoulter** [comment]
Would it be worth following attempting to define a standard json as part of this project. I believe the NRTM software had an associated schema.

**Comment by madarasz** [comment]
Hey, 

TOME support has been on my feature list a long time ago.

It should be possible to decode all the mess (and maybe recode it to the NRTM format that I really like). Exported TOME JSONs would be very welcomed.

The reasons why I haven't spend effort on this so far:
- It's the third most popular Netrunner tournament manager (2nd is ANRTournament, which I also haven't covered yet and ANRTournament also has messy export format.)
- TOME does not require you to enter ID names for players. This usually results in nobody caring and thus the data is unusable for AlwaysBeRunning. Moreover the ID field is free-text, there is a lot of possibility for typos.

As I said, it's on the development queue with low prio. There are other stuff to do currently with greater impact.

I understand that there are a handful of TOs willing to put correct ID names in their TOME tournaments. I would suggest them using the "manual entry" functionality on their concluded tournaments. (there's also a CSV option)

**Comment by ericboivin** [comment]
@lpoulter 
Yes, I do have a few Tome exports, from GNK to Canadian nationals levels. I can share some if you want. I do have access to it too, so we can create fake events to seed data.

@madarasz 
Yeah I agree that it is not the #1 priority, but I don't know if TOs running TOME would submit more their results if there was an easy way to do it! However, Identity recording has been added in the last 6 months! So all that would be missing is maybe a way so that TOs could manually update the list of identities to make sure that the automatic reading could be corrected if they did a mistake.

Could you share a NRTM JSON? I could start looking into the gap between the file formats

**Comment by madarasz** [comment]
@ericboivin - Here are some JSONs for you:
https://alwaysberunning.net/tjsons/411.json
https://alwaysberunning.net/tjsons/414.json
https://alwaysberunning.net/tjsons/330.json (without top-cut)

Currently ID matching in CSV import works with the LIKE '%search_string%' pattern which is looking for substrings. So `Kate` will match `Kate "Mac" McCaffrey: Digital Tinker`. But will not find `Gabriel Santiago: Consummate Professional` for `Gabe`. I'm unsure if complex typo guessing is necessary, maybe just throwing an error that the string cannot be matched is enough.

**Comment by ericboivin** [comment]
Thank you! I will look into that in the next weeks. I might just change my application so that it produce the output in the JSON, and we could have it imported in ABR.
I still haven't found how I will reverse engineer the calculation for SoS based on their JSON! It will be a mess. Do we know the right algorithm for SoS and advanced SoS?

**Comment by gereons** [comment]
@ericboivin Yes, SoS and xSoS calculation is explained in TOMEs help:

> Strength of Schedule
>
>The player with the highest strength of schedule score is ranked highest.
>
>A player's strength of schedule score s is calculated by summing each opponents average score per round a and dividing by the total number of opponents n he has played:
>
>   s = (a[1] + a[2] + ... + a[n]) / n
>  
>
>A player's average score per round a is calculated by summing the number of match points earned p and dividing by the number of matches m the player has played:
>
>   a = (p[1] + p[2] + ... + p[m]) / m
>  
>
>This strength of schedule calculation avoids the issues caused by opponents dropping mid-tournament and first-round byes penalizing the player that receives them.
Extended Strength of Schedule
>
> The player with the highest extended strength of schedule score is ranked highest.
>
> Extended strength of schedule is a new tiebreaker calculated in the same manner as strength of  schedule, but it uses each opponent’s strength of schedule score rather than match points. 

I've implemented this in Objective-C for NRTM, feel free to ask for details.

**Comment by ericboivin** [comment]
Awesome, thanks. I'll figure this out. And what about byes? Are they considered like a played match?

**Comment by gereons** [comment]
I believe they aren't. (I don't have access to the NRTM source right now, will look it up later)

**Comment by GuvBubbs** [comment]
I love when a plan comes together.

**Comment by gereons** [comment]
@ericboivin I remembered correctly: Byes do not count towards SoS/xSoS.

**Comment by Absotively** [comment]
According to [this thread](https://boardgamegeek.com/thread/1456321/tournament-strength-schedule-clarification), byes don't count for SoS, but they _do_ count for the average score per round. That is, they're included when calculating _a_ and omitted when calculating _s_. Has that changed?

**Comment by gereons** [comment]
No, that's still correct. Sorry for being imprecise in my previous comment.

**Comment by GuvBubbs** [comment]
Where are we at with this? The regional season is almost upon us and it would be a good feature to have.

**Comment by ericboivin** [comment]
I am actively following the work done on this project. If this does the job right, I think we could have this as an alternative to NRTM.

[https://github.com/MrHuds0n/golden]

**Comment by GuvBubbs** [comment]
Alternative for sure. I still think NRTM is better.

**Comment by GuvBubbs** [comment]
@ericboivin were you working on this?

**Comment by ericboivin** [comment]
This is what I have so far. I haven't put any work lately though

https://github.com/ericboivin/tome2acoo

**Comment by 5N00P1** [comment]
As you can import TOME into cobr.ai I assume this is not an issue anymore?
